### PR TITLE
[LTO] Add an option to select a pre-link pipeline

### PR DESCRIPTION
**Pull request by igorkudrin**
The option specifies the variant of the optimization pipeline to be used in the LTO backend. The pre-link pipeline can be useful in cases where LTO is used in an intermediate step that combines and optimizes a subset of input files, and the results are used later in the final link.

This implements an idea suggested in https://github.com/llvm/llvm-project/pull/71268#issuecomment-1955086700.

### CODE REVIEW

**Review by aeubanks — APPROVED** [review]

**Comment by nikic** [comment]
Couldn't this also be achieved by using something like `-opt-pipeline='thinlto-pre-link<O3>'`?

**Comment by aeubanks** [comment]
> Couldn't this also be achieved by using something like `-opt-pipeline='thinlto-pre-link<O3>'`?

yes, although I'm not a huge fan of asking people to use the stringly typed interface (even though that's what the C API is to construct passes...)

**Comment by nikic** [comment]
> > Couldn't this also be achieved by using something like `-opt-pipeline='thinlto-pre-link<O3>'`?
> 
> yes, although I'm not a huge fan of asking people to use the stringly typed interface (even though that's what the C API is to construct passes...)

I'd generally agree, but the use-case here sounds very exotic to me, quite far outside the usual LTO workflow. Seems like just the kind of situation that `-opt-pipeline` exists for.

**Comment by igorkudrin** [comment]
> Couldn't this also be achieved by using something like `-opt-pipeline='thinlto-pre-link<O3>'`?

What if the input set contains not only ThinLTO files but also regular ones?

File: llvm/include/llvm/LTO/Config.h [hunk 1]
```diff
@@ -63,6 +63,9 @@ struct Config {
   /// Use the standard optimization pipeline.
   bool UseDefaultPipeline = false;
 
+  /// Use pre-link pipeline.
```
[the remark this answers is not in the document]
Comment (igorkudrin) [reply llvm/include/llvm/LTO/Config.h]: Thanks!

**Comment by igorkudrin** [comment]
Thank you very much. It looks like the ThinLTO pre-link pipeline is indeed working well for our needs.

**Pull request closed without merging**